### ISSUE TITLE
Poetry after 1.1.9 fails to recognize conda environments as valid

### ISSUE THREAD

**Issue by thurse93** [python-poetry/poetry#4566]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: Windows 10 20H2
- **Poetry version**: 1.1.9

## Issue
Since poetry version 1.1.9 poetry does not recognize valid virtual environments created by conda.

We use conda (via tox-conda) to test our libraries with different versions of python (since we're in a corporate network disconnected from the internet we have to rely on repositories for the python interpreters). Minimal tox example to reproduce the problem:

tox.ini
```
[tox]
requires = tox-conda
envlist = debug

[testenv]

[testenv:debug]
deps =
    poetry == 1.1.8
commands =
    python -m poetry env info
```
Produces something like:
```
debug run-test: commands[0] | python -m poetry env info

Virtualenv
Python:         3.7.11
Implementation: CPython
Path:           C:\path\to\project\.tox\debug
Valid:          True
```
Whereas
```
[tox]
requires = tox-conda
envlist = debug

[testenv]

[testenv:debug]
deps =
    poetry == 1.1.9
commands =
    python -m poetry env info
```
Leads to a broken environment (in the eyes of poetry):
```
debug run-test: commands[0] | python -m poetry env info

Virtualenv
Python:         3.7.11
Implementation: CPython
Path:           C:\path\to\project\.tox\debug
Valid:          False
```

We think this is because `conda create` does not put the python executables inside the Scripts directory like virtualenv does. Instead, they are placed in the root directory of the virtualenv folder.

If we understood https://github.com/python-poetry/poetry/pull/4507 right, poetry 1.1.9 expects all binaries in the Scripts (or bin) Directory of the virtualenv directory.

**Comment by zEdS15B3GCwq** [comment]
I'm also having the same issue. Tested poetry 1.1.9 through 1.1.11 (latest).

My setup is the following:

- Windows 10
- conda 4.10.3 (latest)
- system Python 3.9.9
- poetry installed 1.1.9 ... 1.1.11
- conda environment created with `conda create -n test python=3.9 pip=21.3`

pyproject.toml:

    [tool.poetry]
    name = "test"
    version = "0.1.0"
    description = ""
    authors = ["Your Name <you@example.com>"]

    [tool.poetry.dependencies]
    python = ">=3.9.7,<3.10"

    [tool.poetry.dev-dependencies]

    [build-system]
    requires = ["poetry-core>=1.0.0"]
    build-backend = "poetry.core.masonry.api"

First, I executed `poetry show` just to see the output; at first this just created a virtual env, on the second run, however, poetry reported that the conda venv was broken. Then, I tried installing `pynat` as a test package without additional dependencies, and poetry did as if it actually installed something, but it didn't show up in the virtual environment. Command line log (real path and username were simplified; if there's any discrepancy it doesn't mean that's the reason for the problem):

    (test) C:\test>poetry --version
    Poetry version 1.1.9
    
    (test) C:\test>poetry show
    Creating virtualenv test-oXodMIJs-py3.9 in C:\Users\R\AppData\Local\pypoetry\Cache\virtualenvs
    
    (test) C:\test>poetry show
    The virtual environment found in C:\Users\R\Conda\stun seems to be broken.
    Recreating virtualenv test-oXodMIJs-py3.9 in C:\Users\R\AppData\Local\pypoetry\Cache\virtualenvs\test-oXodMIJs-py3.9
    
    (test) C:\test>poetry add pynat
    The virtual environment found in C:\Users\R\Conda\stun seems to be broken.
    Recreating virtualenv test-oXodMIJs-py3.9 in C:\Users\R\AppData\Local\pypoetry\Cache\virtualenvs\test-oXodMIJs-py3.9
    Using version ^0.6.2 for pynat
    
    Updating dependencies
    Resolving dependencies...
    
    Writing lock file
    
    Package operations: 1 install, 0 updates, 0 removals
    
      • Installing pynat (0.6.2)
    
    (test) C:\test>conda list
    # packages in environment at C:\Users\R\Conda\test:
    #
    # Name                    Version                   Build  Channel
    ca-certificates           2021.10.26           haa95532_2
    certifi                   2021.10.8        py39haa95532_0
    openssl                   1.1.1l               h2bbff1b_0
    pip                       21.3.1             pyhd8ed1ab_0    conda-forge
    python                    3.9.7                h6244533_1
    setuptools                58.0.4           py39haa95532_0
    sqlite                    3.36.0               h2bbff1b_0
    tzdata                    2021e                hda174b7_0
    vc                        14.2                 h21ff451_1
    vs2015_runtime            14.27.29016          h5e58377_2
    wheel                     0.37.0             pyhd3eb1b0_1
    wincertstore              0.2              py39haa95532_2
    
    (test) C:\test>poetry remove pynat
    The virtual environment found in C:\Users\R\Conda\stun seems to be broken.
    Recreating virtualenv test-oXodMIJs-py3.9 in C:\Users\R\AppData\Local\pypoetry\Cache\virtualenvs\test-oXodMIJs-py3.9
    Updating dependencies
    Resolving dependencies...
    
    Writing lock file
    
    No dependencies to install or update

I got the same result with poetry versions 1.1.9 through 1.1.11, with an additional error in version 1.1.11:

    (test) C:\test>poetry add pynat
    The virtual environment found in C:\Users\R\Conda\test seems to be broken.
    Recreating virtualenv test-oXodMIJs-py3.9 in C:\Users\R\AppData\Local\pypoetry\Cache\virtualenvs\test-oXodMIJs-py3.9
    Using version ^0.6.2 for pynat
    
    Updating dependencies
    Resolving dependencies...
    
    Writing lock file
    
    Package operations: 1 install, 0 updates, 0 removals
    
      • Installing pynat (0.6.2)
    
      ValueError
    
      File \C:\Users\R\AppData\Local\pypoetry\Cache\artifacts\03\af\d1\ee34d3c0147f5299f3405d5b47a5924c3c6d8f730716205b315573e708\pynat-0.6.2.tar.gz does not exist
    
      at ~\AppData\Roaming\pypoetry\venv\lib\site-packages\poetry\core\packages\file_dependency.py:40 in __init__
           36│             except FileNotFoundError:
           37│                 raise ValueError("Directory {} does not exist".format(self._path))
           38│
           39│         if not self._full_path.exists():
        →  40│             raise ValueError("File {} does not exist".format(self._path))
           41│
           42│         if self._full_path.is_dir():
           43│             raise ValueError("{} is a directory, expected a file".format(self._path))
           44│
    
    
    Failed to add packages, reverting the pyproject.toml file to its original content.

The file exists, so I am guessing that the `\` in front of the path is the culprit. Is this an unrelated bug? If so, perhaps it needs its own issue thread.

TL;DR: poetry >=1.1.9 doesn't work with conda environments, and cannot install packages into them.

**Comment by troll-os** [comment]
Same thing here, tried to install with 1.1.11 and when I import my lib in a python shell I get error on imports of libs such as Pandas which obviously are in my pyproject.toml

**Comment by hgupta** [comment]
Using:
- Windows 10
- [Scoop for Windows](https://scoop.sh/)
- Poetry version 1.1.11, installed via `curl`
- Conda (Miniconda) version 4.10.3
- `base` env python version 3.9.5

Steps taken:
- created a new conda env using python version 3.10.0
- activated the new conda env
- created a new poetry project with `virtualenvs.in-project` set to true, which created a `.venv` directory in the project
- `poetry run python -V` shows version `3.9.5` instead of `3.10.0`

I checked the `poetry.bat` file and it uses the command `python3` instead of `python`. When I check the location for both python and python3, they come from the shims' directory of scoop, which is valid.

But when I activate the newly created conda env, it correctly updates the `PATH` but when I checked their respective locations, python comes from the env location but python3 still comes from the shims' directory.

When I checked the conda env directory, I could not find any python3 executable, but only python and pythonw. So no wonder poetry is always picking from the base env.

Is this something that poetry needs to change or should conda fix the python3 issue?

```
C:\Users\harsh\Desktop\hw
λ which python
/d/scoop/shims/python

C:\Users\harsh\Desktop\hw
λ python -V
Python 3.9.5

C:\Users\harsh\Desktop\hw
λ which python3
/d/scoop/shims/python3

C:\Users\harsh\Desktop\hw
λ python3 -V
Python 3.9.5

C:\Users\harsh\Desktop\hw
λ echo %PATH%
D:\cmder_mini\vendor\bin;D:\scoop\apps\git\current\cmd;D:\scoop\apps\miniconda3\current\condabin;D:\cmder_mini\vendor\conemu-maximus5\ConEmu\Scripts;D:\cmder_mini\vendor\conemu-maximus5;D:\cmder_mini\vendor\conemu-maximus5\ConEmu;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\Program Files\NVIDIA Corporation\NVIDIA NvDLISR;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0;C:\WINDOWS\System32\OpenSSH;C:\Users\harsh\.poetry\bin;D:\scoop\apps\nvm\current\nodejs\nodejs;D:\scoop\apps\ruby26\current\bin;D:\scoop\apps\ruby26\current\gems\bin;D:\scoop\apps\miniconda3\current\scripts;D:\scoop\apps\miniconda3\current\Library\bin;D:\cmder_mini;C:\Users\harsh\AppData\Local\Microsoft\WindowsApps;C:\Users\harsh\AppData\Local\Microsoft\WindowsApps;D:\scoop\apps\git\current\mingw64\bin;D:\scoop\apps\git\current\usr\bin;D:\cmder_mini

C:\Users\harsh\Desktop\hw
λ conda activate py310

C:\Users\harsh\Desktop\hw
(py310) λ which python
/d/scoop/apps/miniconda3/current/envs/py310/python

C:\Users\harsh\Desktop\hw
(py310) λ python -V
Python 3.10.0

C:\Users\harsh\Desktop\hw
(py310) λ which python3
/d/scoop/shims/python3    <-- WHAT? WHY?

C:\Users\harsh\Desktop\hw
(py310) λ python3 -V
Python 3.9.5

C:\Users\harsh\Desktop
(py310) λ poetry new my_project
Created package my_project in my_project

C:\Users\harsh\Desktop
(py310) λ cd my_project\

C:\Users\harsh\Desktop\my_project
(py310) λ poetry run python -V
Creating virtualenv my-project in C:\Users\harsh\Desktop\my_project\.venv
Python 3.9.5    <--- WHAT? WHY?

C:\Users\harsh\Desktop\my_project
(py310) λ poetry config --list
cache-dir = "C:\\Users\\harsh\\AppData\\Local\\pypoetry\\Cache"
experimental.new-installer = true
installer.parallel = true
virtualenvs.create = true
virtualenvs.in-project = true
virtualenvs.path = "{cache-dir}\\virtualenvs"  # C:\Users\harsh\AppData\Local\pypoetry\Cache\virtualenvs
```

**Comment by zljubisic** [comment]
I also experienced this problem.

I created a new conda environment, activate it, and installed poetry in it with `pip poetry install`.
Poetry add/remove works with no issues, pyproject.toml is correctly updated, but no packages in conda environment. 

From time to time (for example with `poetry show`) I get:

```
The virtual environment found in C:\Users\...\miniconda3\envs\test_poetry seems to be broken.
Recreating virtualenv test-poetry-Um856anE-py3.9 in C:\Users\...\AppData\Local\pypoetry\Cache\virtualenvs\test-poetry-Um856anE-py3.9
```

For example, I installed uvicorn (`poetry add uvicorn`), but no uvicorn in conda environment, nor I can starti it with `poetry run uvicorn`.

Do we have any progress with this issue?

**Comment by zEdS15B3GCwq** [comment]
Issue persists in new release 1.1.12.

**Comment by zljubisic** [comment]
So the only workaround is: "With symlinks for python.exe and python39.dll in the conda environment's scripts folder, poetry works without issues", as it is stated above?
Does it work in windows as well?

**Comment by zEdS15B3GCwq** [comment]
> So the only workaround is: "With symlinks for python.exe and python39.dll in the conda environment's scripts folder, poetry works without issues", as it is stated above? Does it work in windows as well?

Windows is the indicated OS in both the OP's and my posts. In my case, symbolic links work without issues. Of course, it's not optimal that for each environment they need to be created manually.

**Comment by zljubisic** [comment]
Thanks

**Comment by finswimmer** [comment]
OMG, `python` is located under `$CONDA_PREFIX` and `pip` under `$CONDA_PREFIX\Scripts` :facepalm: Let's see how to handle this ...

**Comment by finswimmer** [comment]
I've prepared more or less proof-of-concept PR's. Please be so kind to test it:

Poetry 1.1: https://github.com/python-poetry/poetry/pull/5008
Poetry 1.2: https://github.com/python-poetry/poetry/pull/5007

You can easily install it using pipx, e.g.:

```
pipx install --suffix=@conda11 "poetry @ git+https://github.com/finswimmer/poetry.git@ports/issue-4566-broken-conda"
pipx install --suffix=@conda12 "poetry @ git+https://github.com/finswimmer/poetry.git@issue-4566-broken-conda"
```

And then run by using `poetry@conda11` respectively `poetry@conda12`

**Comment by zEdS15B3GCwq** [comment]
So, I had a quick look at the code in `env.py`, and have a question regarding a possible bug. Sorry, I don't have the time to test it  or dig into it right now; please don't bite my head off if I'm wrong. If you understand the code, this may be an easy question for you.

`env.py > Env > _bin`:
```
    def _bin(self, bin: str) -> str:
        """
        Return path to the given executable.
        """
        if self._is_windows and not bin.endswith(".exe"):
            bin_path = self._bin_dir / (bin + ".exe")
        else:
            bin_path = self._bin_dir / bin

        if not bin_path.exists():
            # On Windows, some executables can be in the base path
            # This is especially true when installing Python with
            # the official installer, where python.exe will be at
            # the root of the env path.
            if self._is_windows:
                if not bin.endswith(".exe"):
                    bin_path = self._bin_dir / (bin + ".exe")  # <<< bug?
                else:
                    bin_path = self._path / bin

                if bin_path.exists():
                    return str(bin_path)

            return bin

        return str(bin_path)
```

The marked line seems to be incorrect. At that point it's already known that `self._bin_dir / (bin + ".exe")` doesn't exist, so perhaps it should be `self._path / (bin + ".exe")`? I haven't checked the definition of `self._path` but my guess is that it points to the root of the _venv_, in which case, is it possible that it's related to the issue at hand?
